libs/langchain-community/src/vectorstores/astradb.ts

```diff
@@ -0,0 +1,293 @@
+   * @returns Promise that resolves with an array of documents and their scores.
+   */
+  async similaritySearchVectorWithScore(
+    query: number[],
+    k: number,
+    filter?: CollectionFilter
+  ): Promise<[Document, number][]> {
+    if (!this.collection) {
+      throw new Error("Must connect to a collection before adding vectors");
+    }
+    
+    const cursor = await this.collection.find(
+      filter ?? {},
+      { 
+        sort: { $vector: query },
+        limit: k,
+        includeSimilarity: true,
+      }
+    );
+
+    const results: [Document, number][] = [];
+
+    await cursor.forEach(async (row: Record<string, unknown>) => {
+      
+      const {
+        $similarity: similarity, 
+        $vector: _vector, 
+        [this.idKey]: _id, 
+        [this.contentKey]: content, 
+        ...metadata
+      } = row;
+
+      const doc = new Document({
+        pageContent: content as string,
+        metadata,
+      });
+
+      results.push([doc, similarity as number]);
+    });
+
+    return results;
+  }
+
+
+  /**
+     * Return documents selected using the maximal marginal relevance.
+     * Maximal marginal relevance optimizes for similarity to the query AND diversity
+     * among selected documents.
+     *
+     * @param {string} query - Text to look up documents similar to.
+     * @param {number} options.k - Number of documents to return.
+     * @param {number} options.fetchK - Number of documents to fetch before passing to the MMR algorithm.
+     * @param {number} options.lambda - Number between 0 and 1 that determines the degree of diversity among the results,
+     *                 where 0 corresponds to maximum diversity and 1 to minimum diversity.
+     * @param {CollectionFilter} options.filter - Optional filter
+     *
+     * @returns {Promise<Document[]>} - List of documents selected by maximal marginal relevance.
+     */
+  async maxMarginalRelevanceSearch(
+    query: string,
+    options: MaxMarginalRelevanceSearchOptions<this["FilterType"]>,
+  ): Promise<Document[]> {
+    if (!this.collection) {
+      throw new Error("Must connect to a collection before adding vectors");
+    }
+
+    const queryEmbedding = await this.embeddings.embedQuery(query);
+    
+    const cursor = await this.collection.find(
+      options.filter ?? {},
+      { 
+        sort: { $vector: queryEmbedding },
+        limit: options.k,
+        includeSimilarity: true,
+      }
+    );
+
+    const results = await cursor.toArray() ?? [];
+    const embeddingList: number[][] = results.map((row) => row.$vector as number[]);
+
+    const mmrIndexes = maximalMarginalRelevance(
+      queryEmbedding,
+      embeddingList,
+      options.lambda,
+      options.k
+    );
+
+    const topMmrMatches = mmrIndexes.map((idx) => results[idx]);
+
+    const docs: Document[] = [];
+    topMmrMatches.forEach((match) => {
+      const {
+        $similarity: _similarity, 
+        $vector: _vector, 
+        [this.idKey]: _id, 
```

Comment:
Why prefix with underscore?

Reply:
This was to show that these values are unused.